retis/src/core/probe/kernel/bpf/include/common.h

```diff
@@ -282,6 +283,8 @@ static __always_inline int chain(struct retis_context *ctx)
 	pass_threshold = get_event_size(event);
 	barrier_var(pass_threshold);
 
+	handle_common_types(ctx, event);
+
```

Comment:
> > I think this might be problematic. Not all functions (and therefore not all probes) that have `struct skb_event_reason` as argument are actually dropping the packet.
> 
> Right.
> 
> > We could very easily end up having duplicate drop events that would be terribly confusing. I do think `DropEvent` must come _only_ from `kfree_skb`. Am I missing a use-case?
> > I can see some kind of generic "tracking type" or "type-driven events" for some interesting types but we should be very clear that it just means "this probe happened to have this type and value in its arguments". IMO, A DropEvent has a very different meaning.
> 
> Drops are quite complex. Bottom line is drop reasons are not linked to a packet effectively being dropped: eg. the kernel could be too old (in case we emulate a reason, but I think we should stop doing so) or the reason might be `SKB_COMSUMED` or `SKB_NOT_DROPPED_YET`. Detecting a packet is being dropped is really about seeing it in the right tracepoint: `skb:kfree_skb`; that's all[1]. The `skb-drop` section was just here to hold drop-related information, not that the packet was effectively dropped.
> 
> Now, I agree this is confusing and not a great UX. The naming of `skb-drop` was pretty bad.
> 
> Then on why retrieving those two "common types". More and more functions in the kernel accept those "reason" enums as parameters and are even returning them. Retrieving those values gives very valuable information, as we can understand better the flow of a packet, and how a function was behaving.
> 

Can you give a concrete example?


> Finally (I think that's important for finding a solution) I wasn't sure about how to model those "common types" in the events. I chose a single section per type (and reused `skb-drop` for this), but that might not be the perfect solution.
> 
> I think we should improve both situations.
> 
> One possibility: remove the `skb-drop` section and provide a `drop()` helper for events. A bit like the `tracking_id()` helper in the tracking section. That new helper could be used to flag a packet being dropped in the output, and later in Python. It would just be hidden from the raw JSON event (IMO that is fine). In addition the drop reason should be stored in a new section. The `skb-drop` collector can stay, it has value in probing at the right place.
> 

Maybe that's an option but my feeling is that "packets being dropped" is  way more important than "functions happen to pass or return some drop reason". I think the UX should strongly favor the former and try by all means not to get them mixed.

How about:
- We keep the `DropEvent` linked uniquely to kfree_skb. We make it clear in the string representation that there was a packet drop.
- We add a `SkResetEvent` linked uniquely to trace. We make it clear in the string representation that a rst was actually sent.
- We have a `ProbeArgsEvent` or `ProbeMetadata` or `ProbeInfo` (something like that) that gives information about the probe itself, its arguments and return code for known types. We could support native C types and some well-known structs or enums. Here we can add skb_drop_reason but the string representation should be super-clear this is an argument of the function, e.g: 

```
 [k] [tcp_rcv_state_process] skb ffff8e8ec1ab3300
   probe info: arg0(struct sock*) = 0x123445, arg1(struct skb*) = 0xffff8e8ec1ab3300, ret(enum skb_drop_reason) = SKB_DROP_REASON_TCP_INVALID_SEQUENCE
   192.168.1.137 > 192.168.1.138 ttl 64 tos 0x0 id 49968 off ... rest of skb section ...
   [... other sections ...]

 [k] [kfree_skb_reason] skb ffff8e8ec1ab3300
   192.168.1.137 > 192.168.1.138 ttl 64 tos 0x0 id 49968 off ... rest of skb section
    drop (reason TCP_OLD_SEQUENCE) 
```


> Two additional comments:
> 
>     * I forgot to remove the `skb-drop` section generation from the `skb-drop` collector. It is now handled by the core.
> 
>     * We can split the generic types part from the rest of the series, so we can merge the initial rework.
> 

I definitely think we should as the rest of the PR is almost ready to get merged and can unblock future work (in fact python series would need an interesting  rebase).

> 
> [1] It's in fact even more complex, there are different definitions of what is a "drop". Eg. run `retis -p generic collect -f icmp --cmd 'ping 1.1.1.1'` and watch packets being "drops" :)

I see no drops here, but in any case, even if they are, this is the kernel reporting drops, we cannot do better than the kernel. But we should do worse (i.e: making non-drops feel like drops).